foolbox/attacks/gradient.py

```diff
@@ -119,25 +142,53 @@ def __call__(self, input_or_adv, label=None, unpack=True,
         if not a.has_gradient():
             return
 
+        if stop_early is None:
+            stop_early = bin_search
+        assert (stop_early, bin_search) != (False, True), "When bin_search "\
+            "is true, stop_early must be set to true as well"
+
         image = a.original_image
         min_, max_ = a.bounds()
 
         if not isinstance(epsilons, Iterable):
             assert isinstance(epsilons, int)
-            epsilons = np.linspace(0, 1 / steps, num=epsilons + 1)[1:]
-
-        for epsilon in epsilons:
-            perturbed = image
-
-            for _ in range(steps):
-                gradient = a.gradient(perturbed)
-                gradient_norm = np.sqrt(np.mean(np.square(gradient)))
-                gradient = gradient / (gradient_norm + 1e-8) * (max_ - min_)
-
-                perturbed = perturbed + gradient * epsilon
-                perturbed = np.clip(perturbed, min_, max_)
+            decrease_if_first = True
+            epsilons = np.linspace(0,
+                                   max_epsilon / steps,
+                                   num=epsilons + 1)[1:]
+        else:
+            decrease_if_first = False
 
-                a.predictions(perturbed)
-                # we don't return early if an adversarial was found
-                # because there might be a different epsilon
-                # and/or step that results in a better adversarial
+        for _ in range(2):
+            if bin_search:
+                iterator = BinarySearchIterator(epsilons)
+            else:
+                iterator = LinearSearchIterator(epsilons, stop_early)
+
+            # TODO: add 'make smaller' option if epsilon to small
+            for i, epsilon in iterator:
+                perturbed = image
+
+                for _ in range(steps):
+                    gradient = a.gradient(perturbed)
+                    gradient_norm = np.sqrt(np.mean(np.square(gradient)))
+                    gradient = gradient / (gradient_norm + 1e-8) * (
+                        max_ - min_)
+
+                    perturbed = perturbed + gradient * epsilon
+                    perturbed = np.clip(perturbed, min_, max_)
+
+                    _, is_adversarial = a.predictions(perturbed)
+
+                iterator.is_adversarial = is_adversarial
```

Comment:
using a generator and its send method might be more pythonic, but I think it's fine ;-)